### PR TITLE
TIG-1209 (Part 1): Clean up DocumentGenerator API

### PR DESCRIPTION
**Pull request by rtimmons**
There's not a lot of internal change within value generators here, just solidifying the public API and not exposing variant types outside of the `v1` namespace.

The new API is simple:

- A `*Generator` gets constructed from a `YAML::Node` and a RNG.
- Once constructed invoke it as a function via `operator()` to produce the output type.

I don't know if we have a use-case to expose anything other than `DocumentGenerator` but we can add that in when/if needed.

This is *definitely* not up to a satisfactory level of documentation, testing, and tidyness but wanted to keep the diff down.

Next round of changes will add this in and I'm also going to try to eliminate the `Value` type. Further PRs should be limited entirely to the `value_generators` CMake module.

(cc @dalyd you may find this interesting feel free to tag yourself as required if you want)

### CODE REVIEW

**Comment by visemet** [comment]
> The new API is simple:
> 
> * A `*Generator` gets constructed from a `YAML::Node` and a RNG.

What's the benefit of making the `genny::DefaultRandom` an argument to the constructor? I think it isn't necessary at YAML-parse and is would get in the way of writing `YAML::convert<>::decode()` functions because it isn't available statically.

File: src/value_generators/src/DocumentGenerator.cpp [hunk 1]
```diff
@@ -488,7 +510,11 @@ UniqueExpression RandomStringExpression::parse(YAML::Node node) {
         }
     }
 
-    return std::make_unique<RandomStringExpression>(std::move(length), std::move(alphabet));
+    UniqueTypedExpression<IntegerValueType> lengthT =
+        std::make_unique<TypedExpression<IntegerValueType>>(std::move(length));
+
+    // TODO: does this want to be `optional`?
```
Comment (visemet) [inline src/value_generators/src/DocumentGenerator.cpp]: What is "this"?

File: src/value_generators/include/value_generators/DocumentGenerator.hpp [hunk 2]
```diff
@@ -182,18 +246,45 @@ class Expression {
     virtual Value evaluate(genny::DefaultRandom& rng) const = 0;
 };
 
+template <class T>
+class TypedExpression {
+    UniqueExpression _expression;
```
Comment (visemet) [inline src/value_generators/include/value_generators/DocumentGenerator.hpp]: [style] Explicitly declare a private: section and move these lines after the public: section?

File: src/value_generators/test/expression_test.cpp [hunk 3]
```diff
@@ -230,60 +248,67 @@ null
...
-        REQUIRE_THROWS_AS(Expression::parseOperand(yaml), InvalidValueGeneratorSyntax);
+        REQUIRE_THROWS_AS(Expression::parseOperand(yaml, rng), InvalidValueGeneratorSyntax);
     }
 }
 
 TEST_CASE("Expression parsing with ConstantExpression::parse") {
     genny::DefaultRandom rng{};
     rng.seed(269849313357703264LL);
 
+    SECTION("recursive application caught at setup time") {
```
Comment (visemet) [inline src/value_generators/test/expression_test.cpp]: > > Is this so much of being recursive as just being the wrong type?
> 
> This is proving that the type-checking is done recursively at ctor time not at runtime. This didn't used to fail until `evaluate()` was called.

I feel like "type errors caught at parse time" would be a more understandable name for what this is demonstrating.

File: src/value_generators/include/value_generators/DocumentGenerator.hpp [hunk 4]
```diff
@@ -134,6 +138,64 @@ class Expression;
...
+
+struct IntegerValueType {
+    using OutputType = int64_t;
+    constexpr static ValueType valueType() {
+        return ValueType::Integer;
+    }
+    constexpr static OutputType convert(const Value& value) {
+        auto out = value.tryAsInt64();
+        if (!out) {
+            throw InvalidValueGeneratorSyntax("Unknown");
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: Good call

File: src/value_generators/include/value_generators/DocumentGenerator.hpp [hunk 5]
```diff
@@ -349,15 +457,39 @@ class FastRandomStringExpression : public Expression {
...
 
 private:
-    const UniqueExpression _length;
+    const UniqueTypedExpression<IntegerValueType> _length;
+};
+
+}  // namespace v1
+
+class DocumentGenerator {
+    using UPtr = v1::UniqueTypedExpression<v1::DocumentValueType>;
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: Done - I'll do a thorough audit of all the code ordering and `CONTRIBUTING.md` conventions prior to merging the second pass PR.

File: src/value_generators/src/DocumentGenerator.cpp [hunk 6]
```diff
@@ -250,27 +256,30 @@ UniqueExpression ConstantExpression::parse(YAML::Node node) {
...
-            return std::make_unique<ConstantExpression>(Value{node.as<int64_t>()});
+            return std::make_unique<ConstantExpression>(Value{node.as<int64_t>()},
+                                                        ValueType::Integer);
         } catch (const YAML::BadConversion& e) {
         }
 
         try {
-            return std::make_unique<ConstantExpression>(Value{node.as<double>()});
+            return std::make_unique<ConstantExpression>(Value{node.as<double>()},
+                                                        ValueType::Integer);
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/src/DocumentGenerator.cpp]: Added a `Double` type

File: src/value_generators/src/DocumentGenerator.cpp (hunk 1 above)
Comment (rtimmons) [reply src/value_generators/src/DocumentGenerator.cpp]: lol can't remember my thought process - just removed that todo

**Comment by rtimmons** [comment]
> What's the benefit of making the genny::DefaultRandom an argument to the constructor? I think it isn't necessary at YAML-parse (only at evaluate() time with generating random value) and would get in the way of writing YAML::convert<>::decode() functions because it isn't available statically.

The goal is to make it so Actors don't have to manage the `DefaultRandom` themselves and we may even be able to move it into `v1`. Just make it a `thread_local` in `ActorContext` and use it when calling `ActorContext::get<DocumentGenerator>()`.

**Comment by rtimmons** [comment]
Incorporated feedback and broke out `DoubleValueType` for the places where we need a `double` instead of an `int`.

To keep the next diff small, I tried moving almost all of the impl into the cpp file which *mostly* worked but self-testing of `Expression::parse*` broke and that testing will be valuable to keep around when killing `Value` so I left it as-is for now.

Also cool: the CI test caught me breaking the `create-new-actor.sh` script and requiring doubles for `possion:mean` values in workload yamls. We're doing things right!

I scheduled a patch-build [here](https://evergreen.mongodb.com/version/5c8acf1e0ae6064f431085d8). Assuming that comes back clean I'll merge tomorrow morning.

File: src/value_generators/include/value_generators/DocumentGenerator.hpp [hunk 7]
```diff
@@ -176,15 +178,31 @@ struct IntegerValueType {
...
     }
 };
 
+struct DoubleValueType {
+    using OutputType = double;
+    constexpr static ValueType valueType() {
+        return ValueType::Double;
+    }
+    static OutputType convert(const Value& value) {
+        auto maybeInt = value.tryAsInt64();
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: Ya this is somewhat temporary. I'm going to add a `DoubleGenerator` type and the `ConstantDoubleGenerator` impl will do `Node::as<double>()`.

> not all int64_ts can be exactly represented...

IMHO I'd leave this up to the built-in behavior of `Node::as<double>()`. If it encounters something that can't be converted (e.g. an unreasonably large number of digits) it throws a `bad conversion` exception.

And fwiw on my machine `std::numeric_limits<double>::max()` reports as `1.79769e+308`.

[the remark this answers is not in the document]
Comment (visemet) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: > IMHO I'd leave this up to the built-in behavior of `Node::as<double>()`. If it encounters something that can't be converted (e.g. an unreasonably large number of digits) it throws a `bad conversion` exception.

But you aren't doing that. You're having yaml-cpp convert the value to an `int64_t` and then asking C++ to convert that value to a `double`. C++ isn't going to do you any favors and tell you whether that conversion incurs a loss of precision.

> And fwiw on my machine `std::numeric_limits<double>::max()` reports as `1.79769e+308`.

I'm not sure how that's relevant. [`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) is the relevant value and isn't going to different between machines.

[the remark this answers is not in the document]
Comment (visemet) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: Is there a reason why reverting to say `const double _p` in `BinomialIntExpression`, `NegativeBinomialIntExpression`, and `GeometricIntExpression`, as well as `const double _mean` in `PoissonIntExpression` is undesirable?

[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: `Number.MAX_SAFE_INTEGER` is for javascript, ya? Fwiw I was able to get the cpp yaml parser to spit out a `long long` that was over 100 digits long in base10.

But I think this whole thing is all for naught: I can't think of a reason why we'd want to read yaml-borne values as int32 and not just always use int64. Users don't explicitly specify a type so we just implicitly try it as a 32 or a 64 if necessary. It seems less surprising to always use int64, and it makes downstream concerns a lot easier (especially in testing) to be able to assume that all numbers are either `double`s or `int64`s. Please push back if you disagree - but I may punt supporting `int32`s to a separate ticket where we can figure out a more explicit syntax (maybe similar to how the shell and friends do `{$numberLong:}` etc?)

[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: > Is there a reason why reverting to say const double _p in BinomialIntExpression, NegativeBinomialIntExpression, and GeometricIntExpression, as well as const double _mean in PoissonIntExpression is undesirable?

As opposed to the change to use a generator for the values? It feels "right" to have everything be applied as a template similar to how `RandomString.length` is templated etc. Plus I think it'll make later changes clearer to have everything be consistently a typed-generator.

[the remark this answers is not in the document]
Comment (visemet) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: > `Number.MAX_SAFE_INTEGER` is for javascript, ya?

It's all the same IEEE 754 floating-point value representation.

> Fwiw I was able to get the cpp yaml parser to spit out a long long that was over 100 digits long in base10.

That doesn't make any sense to me because `log(1e100) / log(2) ~= 332`. There aren't enough bits in a 64-bit integral value.

> I can't think of a reason why we'd want to read yaml-borne values as int32 and not just always use int64.

BSON has both `int32_t` and `int64_t` types, which may have different performance characteristics given their different widths. I agree about eventually needing to offer syntax to explicitly specify `int64_t` (e.g. an 'L' suffix) but I don't see why we're throwing away support for `int32_t`.

> As opposed to the change to use a generator for the values? It feels "right" to have everything be applied as a template similar to how `RandomString.length` is templated etc. Plus I think it'll make later changes clearer to have everything be consistently a typed-generator.

Until there is a non-`ConstantExpression` way to generate `double` values, I don't think it is worth it adding the extra layer. We can defer it until we have such an expression and are forced to have a better story around conversions between numeric types.

[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: > I don't see why we're throwing away support for int32_t.

We narrow values into int32 if they fit otherwise we leave them as int32s. This is probably not what users expect - I would expect the output type of the DocumentGenerator to always be the same not "the smallest necessary type that could fit the randomly-generated value". E.g. it is there's no way to explicitly generate an int64 with value 0.

I've reverted the change to kill int32 but I think it may come back to haunt us later. If we want value-generators to be strictly typesafe we'd need to introduce int32 and int64 variants of the `RandomIntExpression` classes.

I also got rid of trying to treat Doubles as int64s. I'll add back the `const double`s.

[the remark this answers is not in the document]
Comment (visemet) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: > We narrow values into int32 if they fit otherwise we leave them as int32s. This is probably not what users expect - I would expect the output type of the DocumentGenerator to always be the same not "the smallest necessary type that could fit the randomly-generated value".

There aren't any randomly generated `int32_t`s - there are only randomly generated `int64_t`s and `std::string`s. `int32_t`s and `double`s can only ever appear as `ConstantExpression`s. I couldn't really speak to what users would expect in this context because there's no explicit syntax (everything would have been `double`s if this were JavaScript). It may not even matter in practice because the server is going widen the type from an `int32_t` to an `int64_t` automatically.

```
> db.mycoll.insert({_id: 0, a: NumberInt(Math.pow(2, 20))})
> db.mycoll.update({_id: 0}, {$mul: {a: NumberInt(Math.pow(2, 20))}})
> db.mycoll.findOne({_id: 0})
{ "_id" : 0, "a" : NumberLong("1099511627776") }
```

What I meant when I said "I don't see why we're throwing away support for `int32_t`" is "I don't see why this pull request would be throwing away support for `int32_t` when its purpose is tackling the syntax for generating values."

> E.g. it is there's no way to explicitly generate an int64 with value 0.

The syntax is laughable awful but you can do

```
{^RandomInt: {min: 0, max: 0}}
```

> I also got rid of trying to treat Doubles as int64s. I'll add back the `const double`s.

Thanks, just ping me again to review when you've made those changes.

File: src/cast_core/src/CrudActor.cpp [hunk 8]
```diff
@@ -364,29 +364,45 @@ struct WriteOperation : public BaseOperation {
 using WriteOpCallback = std::function<std::unique_ptr<WriteOperation>(
     YAML::Node, bool, mongocxx::collection, genny::DefaultRandom&, metrics::Operation)>;
 
+namespace {
+
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/cast_core/src/CrudActor.cpp]: This Actor wants to provide a better error message if the template isn't provided. Most Actors don't have such intense configuration so catching mis-configs are easier.

File: src/value_generators/include/value_generators/DocumentGenerator.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/value_generators/include/value_generators/DocumentGenerator.hpp]: And we're back to `const double`s. Ready for another look whenever you can.

**Comment by rtimmons** [comment]
Thanks for review @dalyd - hopefully the rationale for `createGenerator` makes sense.

> Does passing in the _rng change the behavior/semantics at all? Or does each actor/thread still use the same rng?

It shouldn't, no. There's still the possibility of somebody mistakenly using the same `DefaultRandom` in multiple threads. A later PR will "hide" the `DefaultRandom::create` factory method and call it instead from `ActorContext`.

File: src/workloads/scale/BigUpdate.yml [hunk 9]
```diff
@@ -97,7 +97,7 @@ Actors:
     Threads: 100
     Database: *DB
     CollectionCount: 10  # This is specifically 1% of collections
-    UpdateFilter: {y: {^RandomInt: {distribution: poisson, mean: 100}}}
+    UpdateFilter: {y: {^RandomInt: {distribution: poisson, mean: 100.0}}}
```
Comment (visemet) [inline src/workloads/scale/BigUpdate.yml]: After making the changes from 4f4e69b27d94446e8056c4bc996ff09b233c4148 you should be able to revert the changes to this file as well.
